runtime/shared_common/shrinit.cpp

```diff
@@ -4923,4 +4929,153 @@ isFreeDiskSpaceLow(J9JavaVM *vm, U_64* maxsize)
 	}
 	return ret;
 }
+
+/**
+ * This function generates a key for the GC hints. The key is a string of all the commandline arguments passed to the current JVM
+ * @param[in] vm The current J9JavaVM
+ *
+ * @return A string of all the commandline arguments, NULL if an error occurs.
+ */
+static char*
+generateGCHintsKey(J9JavaVM* vm)
+{
+	JavaVMInitArgs *actualArgs = vm->vmArgsArray->actualVMArgs;
+	UDATA keyLength = 0;
+	UDATA extraLen = 0;
+	UDATA i = 0;
+	UDATA nOptions = vm->vmArgsArray->nOptions;
+	char* key = NULL;
+	bool firstOption = true;
+	PORT_ACCESS_FROM_JAVAVM(vm);
+
+	for (i = 0; i < nOptions; ++i) {
+		char* option = actualArgs->options[i].optionString;
+		if ((NULL != option && strlen(option) > 0) 
+			&& (NULL == strstr(option,"-Dsun.java.launcher.pid="))
+		){
+			keyLength += strlen(actualArgs->options[i].optionString);
+			extraLen += 1;
+		}
+	}
+	if (keyLength == 0) {
+		goto done;
+	}
+	keyLength += extraLen; /* space between options and terminating null char */
+
+	key = (char*)j9mem_allocate_memory(keyLength, J9MEM_CATEGORY_VM);
+	if (NULL == key) {
+		goto done;
+	}
+	memset(key, 0, keyLength);
+	for (i = 0; i < nOptions; ++i) {
+		char* option = actualArgs->options[i].optionString;
+		if ((NULL != option && strlen(option) > 0) 
+			&& (NULL == strstr(option,"sun.java.launcher.pid"))
+		) {
+			if (firstOption) {
+				firstOption = false;
+				j9str_printf(PORTLIB, key, keyLength, "%s%s", key, option);
+			} else {
+				j9str_printf(PORTLIB, key, keyLength, "%s%s%s", key, " ", option);
+			}
+		}
+	}
+done:
+	return key;
+}
+
+/**
+ * Stores the GC hints into the shared classes cache.
+ * @param[in] vmThread  The current thread
+ * @param[in] hintsFlags Extra flags associated with the GC hints
+ * @param[in] heapSize1  The first heap size param that is to be stored into the shared cache
+ * @param[in] heapSize2  The second heap size param that is to be stored into the shared cache
+ * @param[in] forceReplace TRUE Replace the existing GC hints under the same key (the same command line arguments) if there is one already in the shared cache.
+ * 							FALSE Do not replace existing GC hints under the same key.
+ *
+ * @return  The location of the cached GC hints or null
+ */
+const U_8*
+j9shr_storeGCHints(J9VMThread* currentThread, UDATA hintsFlags, UDATA heapSize1, UDATA heapSize2, BOOLEAN forceReplace)
```

Comment:
I think we need a prototype of how it would work to have additional APIs that store and fetch other hints. I'm thinking that instead of j9shr_storeGCHints storing directly to the cache, it should just collect the hints, which are then all stored with a single write to the cache at the end of the startup phase. Similarly for fetching. The cache should fetch all the hints once, perhaps on the first call to get a hint, and then supply the hints from local storage instead of constructing the key and doing a separate read for each separate hint reader.

Reply:
We can do this by caching the key and all the start up hints into a new field in J9SharedClassConfig.